ledger/complete/wal/checkpoint_v6_writer.go

```diff
@@ -0,0 +1,529 @@
+		Str("first_hash", first.RootHash().String()).
+		Uint64("first_reg_count", first.AllocatedRegCount()).
+		Str("last", last.RootHash().String()).
+		Uint64("last_reg_count", last.AllocatedRegCount()).
+		Msgf("storing checkpoint for %v tries to %v", len(tries), outputDir)
+
+	subtrieRoots := createSubTrieRoots(tries)
+
+	estimatedSubtrieNodeCount := estimateSubtrieNodeCount(tries)
+
+	subTrieRootIndices, subTriesNodeCount, subTrieChecksums, err := storeSubTrieConcurrently(
+		subtrieRoots,
+		estimatedSubtrieNodeCount,
+		outputDir,
+		outputFile,
+		logger,
+	)
+	if err != nil {
+		return fmt.Errorf("could not store sub trie: %w", err)
+	}
+
+	logger.Info().Msgf("subtrie have been stored. sub trie node count: %v", subTriesNodeCount)
+
+	topTrieChecksum, err := storeTopLevelNodesAndTrieRoots(
+		tries, subTrieRootIndices, subTriesNodeCount, outputDir, outputFile, logger)
+	if err != nil {
+		return fmt.Errorf("could not store top level tries: %w", err)
+	}
+
+	err = storeCheckpointSummary(subTrieChecksums, topTrieChecksum, outputDir, outputFile, logger)
+	if err != nil {
+		return fmt.Errorf("could not store checkpoint summary: %w", err)
+	}
+
+	logger.Info().Msgf("checkpoint file has been successfully stored at %v/%v", outputDir, outputFile)
+
+	return nil
+}
+
+// 		1. version
+//		2. checksum of each part file (17 in total)
+// 		3. checksum of the main file itself
+func storeCheckpointSummary(
+	subTrieChecksums []uint32,
+	topTrieChecksum uint32,
+	outputDir string,
+	outputFile string,
+	logger *zerolog.Logger,
+) error {
+	closable, err := createWriterForCheckpointSummary(outputDir, outputFile, logger)
+	if err != nil {
+		return fmt.Errorf("could not store checkpoint summary: %w", err)
+	}
+	defer func() {
+		closeErr := closable.Close()
+		// Return close error if there isn't any prior error to return.
+		if err == nil {
+			err = closeErr
+		}
+	}()
+
+	// write version
+	version := encodeVersion(MagicBytes, VersionV6)
+	writer := NewCRC32Writer(closable)
+	_, err = writer.Write(version)
+	if err != nil {
+		return fmt.Errorf("cannot write checksum summary header: %w", err)
+	}
+
+	// TODO: write checksums
+
+	// write checksum to the end of the file
+	checksum := writer.Crc32()
+	_, err = writer.Write(encodeCRC32Sum(checksum))
+	if err != nil {
+		return fmt.Errorf("cannot write CRC32 checksum to checkpoint summary: %w", err)
+	}
+	return nil
+}
+
+func createWriterForCheckpointSummary(outputDir string, outputFile string, logger *zerolog.Logger) (io.WriteCloser, error) {
+	fullPath := filePathHeader(outputDir, outputFile)
+	if utilsio.FileExists(fullPath) {
+		return nil, fmt.Errorf("checkpoint file already exists at %v", fullPath)
+	}
+
+	return createClosableWriter(outputDir, logger, outputFile, fullPath)
+}
+
+// 17th part file contains:
+// 1. checkpoint version TODO
+// 2. checkpoint file part index TODO
+// 3. subtrieNodeCount TODO
+// 4. top level nodes
+// 5. trie roots
+// 6. node count
+// 7. trie count
+// 6. checksum
```

Comment:
you might need to update this part and remove TODOs

Reply:
TODOs are done and cleaned in a different PR